addons/bus/websocket.py

```diff
@@ -571,6 +598,41 @@ def _on_server_stop(cls):
             if websocket.state is ConnectionState.OPEN:
                 websocket.disconnect(CloseCode.GOING_AWAY)
 
+    @classmethod
+    def _add_event_callback(cls, event_type, callback):
+        """
+        Register a function to be called upon a websocket lifecycle
+        event. The environment as well as the websocket related to the
+        event will be passed to the callback.
+        """
+        if not isinstance(event_type, LifecycleEvent):
+            raise ValueError('event_type should be a valid LifecycleEvent')
+        cls._event_callbacks[event_type].add(callback)
+
+    @classmethod
+    def _clear_event_callbacks(cls):
+        """ Remove all lifecycle event callbacks """
+        cls._event_callbacks.clear()
+
+    def _trigger_lifecycle_event(self, event_type):
+        """
+        Trigger a lifecycle event that is, call every function
+        registered for this event type. Every callback is given both the
+        environment and the related websocket.
+        """
+        registry = Registry(self._session.db)
+        with closing(registry.cursor()) as cr:
+            env = api.Environment(cr, self._session.uid, self._session.context)
```

Comment:
We use `closing` in `odoo.http.Request._serve_db` because it is `odoo.service.model.retrying` that is responsible of committing the transaction.

Using `closing` you effectively call `cr.close` at the end of the with-block which rollback the current transaction and close the underlying connection (well it puts it back in a pool in reality).

You should either:

* Don't use not `closing` nor `retrying`, simply `with registry.cursor() as cr`. This makes you vulnerable to serialization failures. Note that it wouldn't crash thanks to the `except Exception` but the cursor would be unusable for further processing. I don't not suggest this solution.
* Leave the `closing` in place and wrap the `callback` call with `retrying`. This makes you safe against serialization errors and will make sure the cursor can still be used. Note that because `retrying` does commit, each callback would be done inside his own transaction, I think it is a good idea to do so. I suggest this solution.